src/Emulator/Peripherals/Peripherals/MTD/OpenTitan_FlashController.cs

```diff
@@ -0,0 +1,1185 @@
+            if(workQueue.Count == 0){
+                if(opState == OperationState.DONE){
+                    opStatusRegisterDoneFlag.Value = true;
+                    interuptStatusRegisterOpDoneFlag.Value = true;
+                }else if(opState == OperationState.RUNNING){
+                    opState = OperationState.DONE;
+                }
+                return;
+            }
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoWork: workQueue.Count = {0}", workQueue.Count);
+            FlashControllerWorkItem workItem = workQueue.Dequeue();
+            if(workItem.eraseBank){
+                DoEraseBank(workItem.eraseBankNumber);
+            }else if(workItem.eraseDataPage){
+                DoEraseDataPage(workItem.erasePageAddress);
+            }else if(workItem.eraseInfoPage){
+                DoEraseInfoPage(workItem.erasePageAddress);
+            }else if(workItem.readData){
+                DoReadData(workItem.readAddress);
+            }else if(workItem.readInfo){
+                DoReadInfo(workItem.readAddress);
+            }else if(workItem.programData){
+                DoProgramData(workItem.programAddress);
+            }else if(workItem.programInfo){
+                DoProgramInfo(workItem.programAddress);
+            }else{
+                this.Log(
+                    LogLevel.Warning,
+                    "OpenTitan_FlashController/DoWork: Unhandled workItem");
+            }
+        }
+
+        private void DoEraseBank(uint bankNumber){
+            /*
+            ** TODO(julianmb): memory protection should be implemented.
+            ** It currently doesnt have a unit test though.
+            ** MP_BANK_CFG register is of interest
+            */
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoEraseBank: bankNumber = {0}",
+                bankNumber);
+
+            long bytesPerBank = FLASH_PAGES_PER_BANK * FLASH_WORDS_PER_PAGE * FLASH_WORD_SZ;
+            long bankEflashOffset = bankNumber * bytesPerBank;
+
+            for(long i = 0 ; i < bytesPerBank ; i++){
+                flash.WriteByte(bankEflashOffset + i, 0xff);
+            }
+        }
+
+        private void DoEraseDataPage(long pageAddress){
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoEraseDataPage: pageAddress = 0x{0:X}",
+                pageAddress);
+            if(IsOperationAllowed(OperationType.ERASE_DATA_PAGE, pageAddress)){
+                long pageOffset = pageAddress - FLASH_MEM_BASE_ADDR;
+                long bytesPerPage = FLASH_WORDS_PER_PAGE * FLASH_WORD_SZ;
+                for(long i = 0 ; i < bytesPerPage ; i++){
+                    flash.WriteByte(pageOffset + i, 0xff);
+                }
+            }else{
+                interruptStatusRegisterOpErrorFlag.Value |= true;
+                opStatusRegisterErrorFlag.Value |= true;
```

Comment:
Does the peripheral support an IRQ line?

Reply:
It looks like there are a few interrupt lines. The current test (https://github.com/lowRISC/opentitan/blob/d9dc048680d89bd630ffae673103509e28017709/sw/device/tests/flash_ctrl_test.c) and driver (https://github.com/lowRISC/opentitan/blob/9b8fa6d0a8ca53173aa6f0c4b3a3f69cc37d3b5c/sw/device/lib/flash_ctrl.c) dont make use of them.

They currently poll the status or op_status registers